unsafe-fields/src/lib.rs

```diff
@@ -0,0 +1,356 @@
+//!         // INVARIANT: `n` is even.
+//!         #[unsafe]
+//!         n: usize,
+//!     }
+//! }
+//!
+//! impl EvenUsize {
+//!     /// Constructs a new `EvenUsize`.
+//!     ///
+//!     /// Returns `None` if `n` is odd.
+//!     pub fn new(n: usize) -> Option<EvenUsize> {
+//!         if n % 2 != 0 {
+//!             return None;
+//!         }
+//!         // SAFETY: We just confirmed that `n` is even.
+//!         let n = unsafe { Unsafe::new(n) };
+//!         Some(EvenUsize { n })
+//!     }
+//! }
+//! ```
+//!
+//! Attempting to swap unsafe fields of the same type is prevented:
+//!
+//! ```compile_fail,E0308
+//! use unsafe_fields::{unsafe_fields, Unsafe};
+//!
+//! unsafe_fields! {
+//!     /// A range.
+//!     pub struct Range {
+//!         // INVARIANT: `lo <= hi`.
+//!         #[unsafe]
+//!         lo: usize,
+//!         #[unsafe]
+//!         hi: usize,
+//!     }
+//! }
+//!
+//! impl Range {
+//!     pub fn swap(&mut self) {
+//!         // ERROR: Mismatched types
+//!         core::mem::swap(&mut self.lo, &mut self.hi);
+//!     }
+//! }
+//! ```
+//!
+//! # Limitations
+//!
+//! Note that we cannot prevent `Unsafe`s from being swapped between the same
+//! field in instances of the same type:
+//!
+//! ```
+//! use unsafe_fields::{unsafe_fields, Unsafe};
+//!
+//! unsafe_fields! {
+//!     /// A `usize` which is guaranteed to be even.
+//!     pub struct EvenUsize {
+//!         // INVARIANT: `n` is even.
+//!         #[unsafe]
+//!         n: usize,
+//!     }
+//! }
+//!
+//! pub fn swap(a: &mut EvenUsize, b: &mut EvenUsize) {
+//!     core::mem::swap(&mut a.n, &mut b.n);
+//! }
+//! ```
+
+use core::marker::PhantomData;
+
+/// A field with safety invariants.
+///
+/// `Unsafe` should not be named directly - instead, use [`unsafe_fields!`] to
+/// declare a type with unsafe fields.
+///
+/// See the [crate-level documentation](crate) for more information.
+#[repr(transparent)]
+pub struct Unsafe<O: ?Sized, F: ?Sized, const NAME_HASH: u128> {
+    _marker: PhantomData<O>,
+    // INVARIANT: `field` is only modified via public `unsafe` methods. User code is never
+    // invoked implicitly except via public `unsafe` methods.
+    field: F,
+}
+
+// NOTE on design: It may seem counter-intuitive to offer an impl of traits that
+// don't require `unsafe` to call. Unfortunately, this is a fundamental
+// requirement if users want to be able to mark their types as `Copy`. Luckily,
+// we can implement `Copy` (and its unavoidable super-trait, `Clone`) without
+// invoking user code or opening up the possibility of modifying the field. We
+// do this by only implementing `Copy` and `Clone` when `F: Copy`. For `Clone`,
+// the user is still able to provide a manual impl, so this does not
+// fundamentally restrict what behavior can be supported.
+impl<O: ?Sized, F: Copy, const NAME_HASH: u128> Copy for Unsafe<O, F, { NAME_HASH }> {}
+impl<O: ?Sized, F: Copy, const NAME_HASH: u128> Clone for Unsafe<O, F, { NAME_HASH }> {
```

Comment:
Food for thought: With `Immutable`, I think we could provide basically all of the standard trait impls. No need to take action on this in this PR.

Reply:
Ack.